js_modules/dagster-ui/packages/ui-core/src/workspace/WorkspaceContext.tsx

```diff
@@ -59,206 +67,277 @@ export const WorkspaceContext = React.createContext<WorkspaceState>(
+        client,
+        query: LOCATION_WORKSPACE_QUERY,
+        key: `${localCacheIdPrefix}${locationWorkspaceKey(name)}`,
+        version: LOCATION_WORKSPACE_QUERY_VERSION,
+        variables: {
+          name,
+        },
+        bypassCache: true,
+      });
+      const entry = locationData?.workspaceLocationEntryOrError;
+      setLocationsData((locationsData) =>
+        Object.assign({}, locationsData, {
+          [name]: entry,
+        }),
+      );
+      return locationData;
+    },
+    [client, getData, localCacheIdPrefix],
+  );
+
+  const [isRefetching, setIsRefetching] = useState(false);
+
+  const locationsToFetch = useMemo(() => {
+    if (!didLoadCachedData) {
+      return [];
     }
-    allTopLevelResourceDetails {
-      id
-      name
+    if (isRefetching) {
+      return [];
     }
-    ...RepositoryInfoFragment
-  }
-
-  fragment WorkspaceSchedule on Schedule {
-    id
-    cronSchedule
-    executionTimezone
-    mode
-    name
-    pipelineName
-    scheduleState {
-      id
-      selectorId
-      status
+    const toFetch = Object.values(locations).filter((loc) => {
+      const prev = prevLocations.current?.[loc.name];
+      const d = locationsData[loc.name];
+      const entry = d?.__typename === 'WorkspaceLocationEntry' ? d : null;
+      return (
+        prev?.updateTimestamp !== loc.updateTimestamp ||
+        prev?.loadStatus !== loc.loadStatus ||
+        entry?.updatedTimestamp !== loc.updateTimestamp ||
+        entry?.loadStatus !== loc.loadStatus
+      );
+    });
+    prevLocations.current = locations;
+    return toFetch;
+  }, [didLoadCachedData, isRefetching, locations, locationsData]);
+
+  useLayoutEffect(() => {
+    if (!locationsToFetch.length) {
+      return;
     }
-  }
+    setIsRefetching(true);
+    Promise.all(
+      locationsToFetch.map(async (location) => {
+        return await refetchLocation(location.name);
+      }),
+    ).then(() => {
+      setIsRefetching(false);
+    });
+  }, [refetchLocation, locationsToFetch]);
+
+  const locationsRemoved = useMemo(
+    () =>
+      Array.from(
+        new Set([
+          ...Object.values(prevLocations.current).filter((loc) => !locations[loc.name]),
+          ...Object.values(locationsData).filter(
+            (loc): loc is WorkspaceLocationNodeFragment =>
+              loc.__typename === 'WorkspaceLocationEntry' && !locations[loc.name],
+          ),
+        ]),
+      ),
+    [locations, locationsData],
+  );
 
-  fragment WorkspaceSensor on Sensor {
-    id
-    jobOriginId
-    name
-    targets {
-      mode
-      pipelineName
+  useLayoutEffect(() => {
+    if (!locationsRemoved.length) {
+      return;
     }
-    sensorState {
-      id
-      selectorId
-      status
-      ...BasicInstigationStateFragment
+    const copy = {...locationsData};
+    locationsRemoved.forEach((loc) => {
+      delete copy[loc.name];
+      indexedDB.deleteDatabase(`${localCacheIdPrefix}${locationWorkspaceKey(loc.name)}`);
```

Comment:
Hmm, I think making `getData` implement permanent caching on top of `client.query` encourages people to use it as a replacement for client.query and not think about this step, which seems pretty important. 

Maybe we expose this as a function in the same file as `getData` and call it `clearCachedData` or something like that, so that they're paired together in the code?

Reply:
Ah yeah thats a good call